src/NServiceBus.AmazonSQS/MessageDispatcher.cs

```diff
@@ -33,25 +34,167 @@ public MessageDispatcher(TransportConfiguration configuration, IAmazonS3 s3Clien
 
         public async Task Dispatch(TransportOperations outgoingMessages, TransportTransaction transaction, ContextBag context)
         {
-            try
+            List<Task> concurrentDispatchTasks = null;
+            foreach (var dispatchConsistencyGroup in outgoingMessages.UnicastTransportOperations.GroupBy(o => o.RequiredDispatchConsistency))
             {
-                var operations = outgoingMessages.UnicastTransportOperations;
-                var tasks = new Task[operations.Count];
-                for (var i = 0; i < operations.Count; i++)
+                switch (dispatchConsistencyGroup.Key)
+                {
+                    case DispatchConsistency.Isolated:
+                        concurrentDispatchTasks = concurrentDispatchTasks ?? new List<Task>();
+                        concurrentDispatchTasks.Add(DispatchIsolated(dispatchConsistencyGroup));
+                        break;
+                    case DispatchConsistency.Default:
+                        concurrentDispatchTasks = concurrentDispatchTasks ?? new List<Task>();
+                        concurrentDispatchTasks.Add(DispatchBatched(dispatchConsistencyGroup));
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException();
+                }
+
+                try
+                {
+                    await Task.WhenAll(concurrentDispatchTasks).ConfigureAwait(false);
+                }
+                catch (Exception e)
                 {
-                    tasks[i] = Dispatch(operations[i]);
+                    Logger.Error("Exception from Send.", e);
+                    throw;
                 }
+            }
+        }
+
+        Task DispatchIsolated(IEnumerable<UnicastTransportOperation> isolatedTransportOperations)
+        {
+            List<Task> tasks = null;
+            // ReSharper disable once LoopCanBeConvertedToQuery
+            foreach (var operation in isolatedTransportOperations)
+            {
+                tasks = tasks ?? new List<Task>();
+                tasks.Add(Dispatch(operation));
+            }
+
+            return tasks != null ? Task.WhenAll(tasks) : TaskExtensions.Completed;
+        }
 
+        async Task DispatchBatched(IEnumerable<UnicastTransportOperation> toBeBatchedTransportOperations)
+        {
+            List<Task<PreparedMessage>> tasks = null;
+            foreach (var operation in toBeBatchedTransportOperations)
+            {
+                tasks = tasks ?? new List<Task<PreparedMessage>>();
```

Comment:
For my curiosity: is this only to avoid an allocation? Is it cheaper to constantly perform the `if` branching?

Reply:
Yes because if we have nothing to enumerate we don't allocate a list. If else should be cheaper than paying for GC